unit_test/docker_container_test.py

```diff
@@ -21,23 +28,50 @@
     DEFAULT_IMAGE = "quay.io/pypa/manylinux2014_ppc64le:2020-05-17-2f8ac3b"
 elif pm == "s390x":
     DEFAULT_IMAGE = "quay.io/pypa/manylinux2014_s390x:2020-05-17-2f8ac3b"
+else:
+    DEFAULT_IMAGE = ""
+
+
+def basis_container_kwargs():
+    """
+    Parametarize different container engine invocations.
+    """
+    HAVE_DOCKER = bool(shutil.which("docker"))
+    if HAVE_DOCKER:
+        yield {"oci_exe": "docker", "docker_image": DEFAULT_IMAGE}
+    HAVE_PODMAN = bool(shutil.which("podman"))
+    if HAVE_PODMAN:
+        # Basic podman usage
+        yield {"oci_exe": "podman", "docker_image": DEFAULT_IMAGE}
+        # VFS Podman usage (for the podman in docker use-case)
+        home = str(Path("~").expanduser())
+        yield {
+            "oci_exe": "podman",
+            "oci_extra_args_common": f"--cgroup-manager=cgroupfs --storage-driver=vfs --root={home}/.local/share/containers/vfs-storage",
+            "oci_extra_args_create": "--events-backend=file --privileged",
+            "oci_extra_args_start": "--events-backend=file --cgroup-manager=cgroupfs --storage-driver=vfs",
```

Comment:
Could these be set in a `containers.conf` in an appropriate location for you? The CLI arguments take the most precedence and I am a little uncomfortable with vfs being the near default.

Reply:
I'm not familiar with `containers.conf`, it looks like it's a container-level thing and not a cibuildwheel level thing?

What is important to me is that cibuildwheel has an easy way to allow the user to modify _how it_ uses containers to build your wheels. If that involves writing a file to my cwd with options instead of specifying then as environs, that's fine. But I don't want to make any system-level modifications. Everything should work locally.

How would one tell cibuildwheel to use a specific container configuration? Is it something you can pass to each Popen invocation of docker/podman commands? If it was something like:

```
export CIBW_CONTAINER_ENGINE=podman
export CIBW_CONTAINER_CONFIG=path-to-containers.conf
```

that would be fine with me, as long as all of the places where I'm using "_common_args" have an equivalent invocation that respects the given conf file.

Reply:
If you forward `CIBW_CONTAINER_CONFIG` as `CONTAINERS_CONF` for the podman invocation it should do just that, I believe. If not the precedence(so I don't forget) is: `built-in -> /usr/share/containers/ -> /etc/containers/ -> ~/.config/containers/ (rootless) -> invocation`

Right now I'm putting my 'soft' defaults in `/etc/containers` of the container since in rootless podman the user config has higher precedence. If you want to override the storage driver to be vfs, that's done inside the `storage.conf` and the configuration may be overridden as `CONTAINER_STORAGE_CONF`.

Reply:
That seems like a better way to handle those special podman options. I'll have to learn about the syntax and options... I've spent a bit of time looking at [this](https://github.com/containers/common/blob/main/pkg/config/containers.conf) and this might take me some time. If anyone feels inclined to create a containers.json that corresponds to how I've setup a extra arguments I pass to docker/podman create, start, and other "common" commands, that would help a lot when I next have time to work on this.

Reply:
> What is important to me is that cibuildwheel has an easy way to allow the user to modify how it uses containers to build your wheels.

Let's talk about this, because this isn't how cibuildwheel has approached problems like this before. The docker 'backend' has no options exposed to customise (beyond the existing env vars that docker itself reads). Why would users need to customise these options for podman?

Reply:
As mentioned in another reply, running podman inside docker, which is running on a host with stricter permission settings (I'll have to ask to figure out what these are, as I don't manage the gitlab-ci servers I use), need to run podman in "rootless" mode and I think there is some other issue with the overlay file driver, which is why the vfs driver is needed. But again, my understanding of containers is limited, 

If I can determine how to get the CI working with only using existing environment variables that docker reads themselves, I'm perfectly happy to use those instead of relying on cibuildwheel (probably a better design decision anyway, although it would be nice if cibuildwheel had a note about how to configuring containers and then point at the docs to CONTAINERS_CONF and CONTAINERS_STORAGE_CONF). What I'm currently trying to figure out is what those config settings are. They unfortunately, do not correspond exactly to command line options, which is a design decision I always find to be frustrating.